### PR TITLE
server: per-Tree option to force re-signing after interval

### PR DESCRIPTION
**Pull request by daviddrysdale**
For a CT log, RFC 6962 s3.5 specifies that the log should produce a new
STH within the MMD even if there have been no additions to the log.

To allow a CT personality to get the same behavior, add a per-Tree option
to force regeneration of a signed log root after an interval.

This option is implemented in the signer rather than in the personality
because a different user of Trillian-as-a-service may also have the same
requirement that led to RFC 6962 s3.5.

Note that similar functionality was previously removed in commit 0b66fd9
("Remove the code for signing log roots if the last one is too old.")

Addresses issue #575

### CODE REVIEW

**Comment by Martin2112** [comment]
FYI #266 is where I removed it. We might need to revisit this design decision again.

**Comment by daviddrysdale** [comment]
Re-worked to make this a per-`trillian.Tree` option.   From the commit message:

For a CT log, RFC 6962 s3.5 specifies that the log should produce a new
STH within the MMD even if there have been no additions to the log.

To allow a CT personality to get the same behavior, add a per-Tree option
to force regeneration of a signed log root after an interval.

This option is implemented in the signer rather than in the personality
because a different user of Trillian-as-a-service may also have the same
requirement that led to RFC 6962 s3.5.

File: trillian.proto [hunk 1]
```diff
@@ -139,6 +139,10 @@ message Tree {
   // The public key used for verifying tree heads and entry timestamps.
   // Readonly.
   PublicKey public_key = 14;
+
+  // Interval after which a new signed root is produced even if there have been
+  // no submission.  If zero, this behavior is disabled.
+  int64 force_new_root_millis = 15;
```
Comment (codingllama) [inline trillian.proto]: Use `google.protobuf.Duration` and drop the _millis suffix?

Comment (codingllama) [inline trillian.proto]: Maybe `max_root_duration`, where 0 means infinite? "force"-type options are usually non-default, to be used with care (at least in my experience), whereas this is perfectly fine to set.

File: log/sequencer.go [hunk 2]
```diff
@@ -47,6 +47,9 @@ type Sequencer struct {
 	// sequencerGuardWindow is used to ensure entries newer than the guard window will not be
 	// sequenced until they fall outside it. By default there is no guard window.
 	sequencerGuardWindow time.Duration
+	// forceNewRootInterval is used to ensure that a new signed log root is generated after a while,
+	// even if no entries have been added to the log.  Zero duration disables this behavior.
+	forceNewRootInterval time.Duration
```
Comment (codingllama) [inline log/sequencer.go]: Having this is a sequencer attribute means it can go out of sync with the actual tree configuration. This should be a parameter instead.

File: log/sequencer_test.go [hunk 3]
```diff
@@ -281,12 +281,49 @@ func TestSequenceWithNothingQueued(t *testing.T) {
...
 
 	leaves, err := c.sequencer.SequenceBatch(ctx, params.logID, 1)
-	if leaves != 0 {
-		t.Fatalf("Unexpectedly sequenced %d leaves on error", leaves)
+	if leaves != 0 || err != nil {
+		t.Errorf("SequenceBatch()=(%d,%v); want (0,nil)", leaves, err)
 	}
+}
 
+func TestSequenceWithNothingQueuedForceNewRoot(t *testing.T) {
```
Comment (codingllama) [inline log/sequencer_test.go]: Could you join this test and the one above in a single, table-driven test?

File: log/sequencer_test.go [hunk 4]
```diff
@@ -281,12 +281,49 @@ func TestSequenceWithNothingQueued(t *testing.T) {
 	c, ctx := createTestContext(ctrl, params)
 
 	leaves, err := c.sequencer.SequenceBatch(ctx, params.logID, 1)
-	if leaves != 0 {
-		t.Fatalf("Unexpectedly sequenced %d leaves on error", leaves)
+	if leaves != 0 || err != nil {
+		t.Errorf("SequenceBatch()=(%d,%v); want (0,nil)", leaves, err)
```
Comment (codingllama) [inline log/sequencer_test.go]: nit: `s/%d/%v/`. `%v` makes life a lot easier in face of refactors.

Ditto for others.

File: server/sequencer_manager.go [hunk 5]
```diff
@@ -77,6 +77,7 @@ func (s *SequencerManager) ExecutePass(ctx context.Context, logID int64, info *L
 
 	sequencer := log.NewSequencer(hasher, info.TimeSource, s.registry.LogStorage, signer)
 	sequencer.SetGuardWindow(s.guardWindow)
+	sequencer.SetForceNewRootInterval(time.Duration(tree.ForceNewRootMillis * 1000000))
```
Comment (codingllama) [inline server/sequencer_manager.go]: Use `time.Millisecond` instead of 1000000?

File: log/sequencer.go (hunk 2 above)
Comment (codingllama) [inline log/sequencer.go]: It's still misrepresented and at a larger scope than necessary. LogID is a parameter, so this should be too.

File: log/sequencer_test.go (hunk 3 above)
Comment (codingllama) [inline log/sequencer_test.go]: I disagree, both are essentially root duration tests. The first is `duration=0`, this one is `duration>root`. Btw, we should add `duration > 0 && duration < root` and `duration > 0 && duration == root` too.

File: trillian.proto (hunk 1 above)
Comment (daviddrysdale) [inline trillian.proto]: On a quick look, the protobuf `Duration` type doesn't look helpful -- it has separate [seconds and nanoseconds](https://github.com/google/protobuf/blob/master/src/google/protobuf/duration.proto#L103-L117) and so would need non-trivial conversion functions.

File: log/sequencer.go (hunk 2 above)
Comment (daviddrysdale) [inline log/sequencer.go]: The current position was not misrepresented and was consistent with the existing `sequencerGuardWindow` field.  Nevertheless, I have reorganized the pre-existing code to move the fields to be parameters of `SequenceBatch` instead.

File: log/sequencer_test.go (hunk 3 above)
Comment (daviddrysdale) [inline log/sequencer_test.go]: Done; to preserve consistency in the file I converted everything to be table-driven.

However, I do think that this level of change is out of scope for this pull request.

File: trillian.proto (hunk 1 above)
Comment (codingllama) [inline trillian.proto]: Duration is standard for Google APIs. I'd strongly suggest we keep to it. (On a similar case, we should be using `google.protobuf.Timestamp` where appropriate as well.)

There are already conversion functions between `durationpb.Duration` and `time.Duration` under [ptypes](https://godoc.org/github.com/golang/protobuf/ptypes#Duration).

(FWIW, personally, I find the Seconds/Nanos representation a bit strange too.)

File: log/sequencer_test.go (hunk 3 above)
Comment (codingllama) [inline log/sequencer_test.go]: Thanks for changing the entire file, you went above and beyond here. This will go a long way to improve long-term maintainability.

**Comment by codingllama** [comment]
I've just replied to comments for now, I'll take a more careful look asap.

**Comment by rolandshoemaker** [comment]
Was this closed by accident or have you decided on taking a different path?
